### PR TITLE
Add Assets and Basic Orders

### PR DESCRIPTION
**Pull request by gagdiez**
In this commit I added the Class Assets and added as predefined assets
the ones that I use the most. I also made the Class Order, and create
the classes for the following types of orders: Market (simply order);
Limit and TrailingStop. While I briefly tested the classes, I didn't
add such tests in this commit. In the following commits I'll create
more and add them. The three orders are heavily commented and have
examples to help understand them.

### CODE REVIEW

**Comment by gagdiez** [comment]
@sdeslauriers. We don't need to merge this right away, since it doesn't have proper testing. But we can start discussing about the Classes, their Methods and the necessary modifications to other files, such as exchanges/__init__.py

File: exchanges/orders.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+         start following the market price and keep record of the MINIMUM price
+         achieved SINCE this order was placed. If at some point, the market
+         price goes OVER the minimum price by the desired offset, then an
+         Order(BUY, (A0, A1), volume) will be triggered.
+
+       - For the action SELL, and a STRICTLY NEGATIVE offset, this Order will
+         start following the market price and keep record of the MAXIMUM price
+         achieved SINCE this order was placed. If at some point, the market
+         price goes UNDER the maximum price by the desired offset, then an
+         Order(BUY, (A0, A1), volume) will be triggered.
```
Comment (gagdiez) [inline exchanges/orders.py]: SELL

File: exchanges/assets.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# Cryptocurrencies
+Bitcoin = Asset('bitcoin', 'X', 'BTC', 'BTC')
+Ethereum = Asset('ethereum', 'X', 'ETH', 'BTC')
+Litecoin = Asset('litecoin', 'X', 'LTC', 'LTC')
+Ripple = Asset('ripple', 'X', 'XRP', 'XRP')
+Zcoin = Asset('zcoin', 'X', None, 'ZEC')
+
+# Currencies
+Euro = Asset('euro', 'Z', 'EUR', 'EUR')
+Dolar = Asset('dolar', 'Z', 'USD', 'USD')
```
Comment (sdeslauriers) [inline exchanges/assets.py]: Should be Dollar.

File: exchanges/assets.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+class Asset(object):
+
+    def __init__(self, name, atype, iso4217_a3=None, abbreviation=None):
+
+        if iso4217_a3 is None and abbreviation is None:
+            raise ValueError('At least one denomination is needed')
```
Comment (sdeslauriers) [inline exchanges/assets.py]: We will probably never see this error, but the message is not clear to me. Maybe add 'You must provided either an abbreviation or the iso4217_a3 name.'

File: exchanges/assets.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        if atype not in ['X', 'Z']:
+            raise ValueError("atype must be either 'X' or 'Z'")
```
Comment (sdeslauriers) [inline exchanges/assets.py]: Do X and Z have semantic meaning? If so, we could add an enumeration (new in python3) for them or define class constants.

File: exchanges/assets.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+class Asset(object):
+
+    def __init__(self, name, atype, iso4217_a3=None, abbreviation=None):
```
Comment (sdeslauriers) [inline exchanges/assets.py]: Since many of our tools expect asset pairs, it could be nice to have syntactic sugar to define them. For example, we could overload `__add__` to return a tuple. Then you could write
```
print(kraken.lastest_trade(Bitcoin + Dollar))
```

File: exchanges/assets.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    @property
+    def iso4217_a3(self):
+        return self._iso4217_a3
+
+    @property
+    def abbreviation(self):
+        return self._abbreviation
+
+    @property
+    def code(self):
```
Comment (sdeslauriers) [inline exchanges/assets.py]: Because this returns a string version of the currency, this could actually be `__str__`.

File: exchanges/orders.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
+from .assets import Asset
+
+def basic_order_check(action, asset_pair):
+    if action not in ['buy', 'sell']:
```
Comment (sdeslauriers) [inline exchanges/orders.py]: For `buy` and `sell`, we could define an enumeration or class constants.

File: exchanges/orders.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        raise ValueError("Action must be 'buy' or 'sell'")
+
+    asset0, asset1 = asset_pair
+    if not isinstance(asset0, Asset) or not isinstance(asset1, Asset):
+        raise ValueError("Not recognized asset pair")
```
Comment (sdeslauriers) [inline exchanges/orders.py]: This error message is misleading. The problem is not that the pair is not recognized, it is that one of the assets is not an `Asset`. Should be something like 
```
An asset pair must be composed of two instances of Asset, not ({}, {})'.format(class(asset0), class(asset1))
```

Comment (sdeslauriers) [reply exchanges/orders.py]: Also, this deals purely with `Asset`. We could add a convenience function in `exchanges.assets` which you would call here. `validate_asset_pair` for example.

File: exchanges/orders.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+       relationship: volume(A0)*price = volume(A1). Where `price` is dictated
+       by the market
+
+       Examples
+       --------
+       Order(action:BUY, asset_pair:(ETH, BTC), volume:1.5) means:
+       "BUY 1.5 Ethereum using as much Bitcoin as the market dictates"
+
+       Order(action:SELL, asset_pair:(Bitcoin, Ethereum), volume:1.5) means:
+       'SELL 1.5 Ethereum for the amount of Bitcoin the market dictates'"""
```
Comment (sdeslauriers) [inline exchanges/orders.py]: Very clear, but is missing a description of the optional parameters.

File: exchanges/orders.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    """This is the simplest type of order, given an action (i.e. BUY/SELL);
+       an assets_pair (A0, A1) and a volume, it represents:
+
+       BUY/SELL `volume` of A0 using/for some volume of A1, which follows the
+       relationship: volume(A0)*price = volume(A1). Where `price` is dictated
+       by the market
+
+       Examples
+       --------
+       Order(action:BUY, asset_pair:(ETH, BTC), volume:1.5) means:
```
Comment (sdeslauriers) [inline exchanges/orders.py]: I prefer my examples to be valid code, something that can be copypasted to ipython and works. Just my opinion, no need to change it.

File: exchanges/orders.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    @property
+    def volume(self):
+        return self._volume
+
+    @property
+    def leverage(self):
+        return self._leverage
+
+    @property
+    def validate(self):
```
Comment (sdeslauriers) [inline exchanges/orders.py]: Using the verb validate for this is confusing, it sounds like you want to validate the order. Maybe `is_mock`?

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+from setuptools import setup
```
Comment (sdeslauriers) [inline setup.py]: It's nitpicking, but adding the setup file is not inline with the rest of your changes. Could we do this in a different PR?

File: exchanges/orders.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+def basic_order_check(action, asset_pair):
+    if action not in ['buy', 'sell']:
+        raise ValueError("Action must be 'buy' or 'sell'")
+
+    asset0, asset1 = asset_pair
+    if not isinstance(asset0, Asset) or not isinstance(asset1, Asset):
+        raise ValueError("Not recognized asset pair")
+
+
+class Order(object):
```
Comment (sdeslauriers) [inline exchanges/orders.py]: With the current implementation, an order is just data that will be given to the `place_order` method of `Exchange`. This means that the method will have to check the order type (limit, market, ...) and respond correctly. 

As a consequence:

1. if we add an order type, all exchanges will have to be updated to handle the new case. 
2. some exchanges may not implement a certain order type and will crash at run time when given that type.

For 2, I think we can build a system like the currency one where the exchange gives you a list of orders it can handle. For 1, I don't believe there is a way to avoid some manual management, the API of exchanges will be different and we will have to write custom code for each. We just have to make sure this complexity does not bubble up to traders.

**Review by sdeslauriers — COMMENTED** [review]
I like it, but I think we need to brainstorm on how to implement orders.
